File: algorithmAndDataStructure/golang/maximumProductSubarray/maximumProductSubarray.go
package maximumProductSubarray

func min(x int, others ...int) (minValue int) {
	minValue = x
	for _, other := range others {
		if minValue > other {
			minValue = other
		}
	}
	return
}

func max(x int, others ...int) (maxValue int) {
	maxValue = x
	for _, other := range others {
		if maxValue < other {
			maxValue = other
		}
	}
	return
}

func maxProduct(nums []int) int {
	prevMin, prevMax := nums[0], nums[0]
	curMin, curMax := nums[0], nums[0]
	maxValue := nums[0]

	numsLen := len(nums)
	for i := 1; i < numsLen; i++ {
		curMin = min(prevMin*nums[i], prevMax*nums[i], nums[i])
		curMax = max(prevMin*nums[i], prevMax*nums[i], nums[i])
		maxValue = max(maxValue, curMax)
		prevMin, prevMax = curMin, curMax
	}
	return maxValue
}

File: algorithmAndDataStructure/cpp/maximumProductSubarray/maximumProductSubarray.cpp
#include <iostream>
#include <vector>
#include <limits.h>

using namespace std;

int min(int x, int y, int z) {
    return min(x, min(y, z));
}

int max(int x, int y, int z) {
    return max(x, max(y, z));
}

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int prevMin = nums[0], prevMax = nums[0];
        int curMin = nums[0], curMax = nums[0], maxValue=nums[0];
        int numsLen = nums.size();
        for (int i=1; i<numsLen; i++) {
            curMin = min(prevMin*nums[i], prevMax*nums[i], nums[i]);
            curMax = max(prevMin*nums[i], prevMax*nums[i], nums[i]);

            prevMin = curMin, prevMax = curMax;
            maxValue = max(maxValue, curMax);
        }
        return maxValue;
    }
};